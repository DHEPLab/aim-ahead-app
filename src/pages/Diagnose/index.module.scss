.header {
    background: #7097A8;
    color: #303230;
    font-size: 28px;
    font-weight: 400;
    line-height: 60px; 
    padding: 4px 0;
    padding-left: 40px;
}

.container {
    margin: 2rem 28px;

    .index {
        margin: 1rem;
        margin-top: 2rem;

        .circle {
            display: inline-block;
            width: 1.5rem;
            height: 1.5rem;
            line-height: 1.5rem;
            text-align: center;
            border-radius: 50%;
            border: 1px solid #303230;
            margin-right: 4px;
        }

        &.require::after {
            content: "*";
            font-size: 1.2rem;
        } 
    }

    .otherFiled {
        margin: 2rem 0;
    }

    .submit {
        width: 124px;
        line-height: 40px;
        letter-spacing: 0.55px;
        font-weight: 500;
    }
}