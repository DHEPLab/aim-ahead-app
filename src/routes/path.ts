export const path = {
  root: "/",
  signup: "/signup",
  login: "/login",
  resetPassword: "/reset-password/:token",
  case: "/case-review/:caseConfigId",
  answer: "/answer/:caseConfigId",
  forgotPassword: "/forgot-password",
};

export default path;
