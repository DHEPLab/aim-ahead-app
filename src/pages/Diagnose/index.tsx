import React, { ChangeEvent, useLayoutEffect, useMemo, useState } from "react";
import { useNavigate, useParams } from "react-router-dom";
import { Button, Snackbar, TextField } from "@mui/material";
import SyncIcon from "@mui/icons-material/Sync";
import ErrorOutlineIcon from "@mui/icons-material/ErrorOutline";
import { useRequest } from "ahooks";
import { useAtom } from "jotai";
import cls from "classnames";

import { caseAtom, diagnoseAtom } from "../../state";
import Diagnosis, { DiagnosisProps } from "../../components/Diagnosis";
import CaseTitle from "../../components/CaseTitle";
import { saveDiagnose } from "../../services/diagnoseService";
import path from "../../routes/path";

import styles from "./index.module.scss";

type DiagnoseValue = DiagnosisProps["value"];

const DiagnosisCount = 3;

const DEFAULT_DIAGNOSE_VALUE = {
  diagnosis: "",
  rationale: "",
  confidence: 0,
};

export type DiagnoseFormData = {
  diagnose?: DiagnoseValue[];
  other?: string;
};

const Diagnose = () => {
  const { caseConfigId } = useParams() as { caseConfigId: string };
  const nav = useNavigate();
  const [showToast, setShowToast] = useState(false);

  const { loading, runAsync } = useRequest(saveDiagnose, {
    manual: true,
  });

  const [diagnoseState, setDiagnoseState] = useAtom(diagnoseAtom);
  const [caseState] = useAtom(caseAtom);
  const defaultValue = diagnoseState[caseConfigId];

  const [diagnose, setDiagnose] = useState<DiagnoseValue[]>(
    defaultValue?.diagnose ?? (Array.from({ length: DiagnosisCount }).fill(DEFAULT_DIAGNOSE_VALUE) as DiagnoseValue[]),
  );
  const [other, setOther] = useState(defaultValue?.other ?? "");
  const [disable, setDisable] = useState(true);
  const [errorMsg, setErrorMsg] = useState("");

  const diagnoseValuesIsRequired = useMemo(() => {
    return Array.from({ length: DiagnosisCount }).map((_, index) => {
      return index === 0;
    });
  }, []);

  const handelToastOnClose = () => {
    setShowToast(false);
    nav(path.root);
  };

  const handleOnDiagnoseChange = (index: number, value: DiagnoseValue) => {
    const updatedDiagnose = diagnose.reduce((prev, current, currentIdx) => {
      if (currentIdx === index) {
        return [...prev, value];
      }
      return [...prev, current];
    }, [] as DiagnoseValue[]);

    setDiagnose(updatedDiagnose);

    setDiagnoseState({
      ...diagnoseState,
      [caseConfigId]: {
        diagnose: updatedDiagnose,
        other: other,
      },
    });
  };

  const handleOtherChange = (e: ChangeEvent<HTMLTextAreaElement>) => {
    setOther(e.target.value);

    setDiagnoseState({
      ...diagnoseState,
      [caseConfigId]: {
        diagnose,
        other: e.target.value,
      },
    });
  };

  useLayoutEffect(() => {
    const isDisabled = diagnose.some((value = {}, idx) => {
      return Object.values(value).some((field) => {
        return diagnoseValuesIsRequired[idx] && (field === null || field === undefined || field === "");
      });
    });

    setDisable(isDisabled);
  }, [diagnose, setDisable, diagnoseValuesIsRequired]);

  const onSubmit = () => {
    runAsync(caseConfigId, {
      diagnose,
      other,
    })
      .then(() => {
        setShowToast(true);
      })
      .catch((e: Error) => {
        setErrorMsg(e.message);
      });
  };

  return (
    <>
      <div className={styles.header}>Diagnose</div>
      <CaseTitle
        name={caseState.personName}
        case={`Case ${caseState.caseNumber}`}
        suffix={
          <Button variant="contained" color="secondary" onClick={() => nav(-1)}>
            CASE REVIEW
          </Button>
        }
      />
      <div className={styles.container}>
        {diagnose.map((value, i) => {
          const indexStartAt1 = i + 1;
          const required = diagnoseValuesIsRequired[i];

          return (
            <div key={i}>
              <div className={cls(styles.index, { [styles.require]: required })}>
                <span className={styles.circle}>{indexStartAt1}</span>
              </div>
              <Diagnosis
                key={i}
                required={required}
                value={value}
                onChange={(value) => handleOnDiagnoseChange(i, value)}
              />
            </div>
          );
        })}
        <TextField
          fullWidth
          multiline
          rows={4}
          className={styles.otherFiled}
          name="other"
          value={other}
          onChange={handleOtherChange}
          inputProps={{ maxLength: 1000 }}
          label="What other information would have been useful?"
        />
        {errorMsg && (
          <div className={styles.errorContainer}>
            <ErrorOutlineIcon />
            <span>{errorMsg}</span>
          </div>
        )}
        <Button
          className={styles.submit}
          disabled={disable || loading}
          variant="contained"
          onClick={onSubmit}
          endIcon={loading && <SyncIcon className={styles.spin} />}
        >
          Submit
        </Button>
        <Snackbar
          anchorOrigin={{ vertical: "bottom", horizontal: "center" }}
          open={showToast}
          onClose={handelToastOnClose}
          autoHideDuration={500}
          message="Task is submitted."
        />
      </div>
    </>
  );
};

export default Diagnose;
