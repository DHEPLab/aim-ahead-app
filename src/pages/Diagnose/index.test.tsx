import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter, Route, Routes } from "react-router-dom";

import Diagnose from ".";
import path from "../../routes/path";

it("Diagnose page snapshot test", () => {
  expect(
    render(
      <MemoryRouter initialEntries={["/diagnose/1"]}>
        <Routes>
          <Route path={path.diagnose} element={<Diagnose />} />
        </Routes>
      </MemoryRouter>,
    ),
  ).toMatchSnapshot();
});
