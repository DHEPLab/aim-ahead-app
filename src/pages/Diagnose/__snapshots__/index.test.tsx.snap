// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Diagnose page snapshot test 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="header"
      >
        Diagnose
      </div>
      <div
        class="container"
      >
        <div>
          <div
            class="name"
          />
          <div
            class="case"
          >
            Case undefined
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary css-zcbmsk-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          CASE REVIEW
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
      <div
        class="container"
      >
        <div />
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary submit css-sghohy-MuiButtonBase-root-MuiButton-root"
          data-testid="diagnose-submit-btn"
          disabled=""
          id="diagnose-submit-btn"
          tabindex="-1"
          type="button"
        >
          Submit
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="header"
    >
      Diagnose
    </div>
    <div
      class="container"
    >
      <div>
        <div
          class="name"
        />
        <div
          class="case"
        >
          Case undefined
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary css-zcbmsk-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        CASE REVIEW
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
    <div
      class="container"
    >
      <div />
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary submit css-sghohy-MuiButtonBase-root-MuiButton-root"
        data-testid="diagnose-submit-btn"
        disabled=""
        id="diagnose-submit-btn"
        tabindex="-1"
        type="button"
      >
        Submit
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
